File: 卡码/代码随想录/Go/动态规划/02-背包问题/01-01背包/LeetCode_0416.go
// 416. 分割等和子集
package main

import "math"

func canPartition(nums []int) bool {
	sum:=0
	for _,num:=range nums {
		sum+=num
	}
	if sum%2==1 {
		return false
	}
	target:=sum/2
	dp:=make([]int,10001)
	for i:=0;i<len(nums);i++ {
		for j:=target;j>=nums[i];j-- {
			dp[j]=int(math.Max(float64(dp[j]),float64(dp[j-nums[i]]+nums[i])))
		}
	}
	if dp[target]==target {
		return true
	}
	return false
}

File: 卡码/代码随想录/JS/动态规划/02-背包问题/01-01背包/LeetCode_0416.js
//416. 分割等和子集
const canPartition = function(nums) {
	let sum=0;
	for(let num of nums) sum+=num;
	if(sum%2==1) return false;
	const target=Math.floor(sum/2);
	const dp=new Array(10001).fill(0);
	for(let i=0;i<nums.length;i++){
		for(let j=target;j>=nums[i];j--){
			dp[j]=Math.max(dp[j],dp[j-nums[i]]+nums[i]);
		}
	}
	if(target==dp[target]) return true;
	return false;
};